includes/class-wc-gateway-dintero-hp.php

```diff
@@ -1233,130 +1220,240 @@ private function capture( $order, $transaction = null ) {
 				$transaction = self::_adapter()->get_transaction($transaction_id);
 			}
 
+			if (isset($transaction['error'])) {
+				$order->add_order_note(__('Could not capture transaction: ' . $transaction['message']));
+				$order->set_status( 'on-hold' );
+				$order->save();
+				$order->save_meta_data();
+				return false;
+			}
+
+			if ($this->extract('status', $transaction) === 'CAPTURED') {
+				$order->add_order_note('Payment captured via Dintero, already captured from before.');
+				$order->save_meta_data();
+				return false;
+			};
+
+			if ($this->extract('status', $transaction) !== 'AUTHORIZED') {
+				$order->add_order_note(__(
+					sprintf(
+						'Could not capture transaction: Transaction status is wrong (%s).',
+						$this->extract('status', $transaction),
+					)
+				));
+				$order->set_status( 'on-hold' );
+				$order->save();
+				$order->save_meta_data();
+				return false;
+			}
+
 			$order_total_amount = absint( strval( floatval( $order->get_total() ) * 100 ) );
 
-			if ( array_key_exists( 'status', $transaction ) &&
-				 array_key_exists( 'amount', $transaction ) &&
-				 'AUTHORIZED' === $transaction['status'] &&
-				 $transaction['amount'] >= $order_total_amount ) {
-
-				$items = array();
-
-				$counter = 0;
-				foreach ( $order->get_items() as $order_item ) {
-					$counter ++;
-					$line_id                = strval( $counter );
-					$item_total_amount      = absint( strval( floatval( $order_item->get_total() ) * 100 ) );
-					$item_tax_amount        = absint( strval( floatval( $order_item->get_total_tax() ) * 100 ) );
-					$item_line_total_amount = absint( strval( floatval( $order->get_line_total( $order_item,
-							true ) ) * 100 ) );
-					$item_tax_percentage    = $item_total_amount ? ( round( ( $item_tax_amount / $item_total_amount ),
-							2 ) * 100 ) : 0;
-					if ( $order_item['variation_id'] ) {
-						$product = wc_get_product( $order_item['variation_id'] );
-					} else {
-						$product = wc_get_product( $order_item['product_id'] );
-					}
-					$item_reference = $product->get_id();
-
-					$productId =  substr( (string) $item_reference, 0, 64 );
-					$item                   = array(
-						'id'          => $productId,
-						'description' => $order_item->get_name(),
-						'quantity'    => $order_item->get_quantity(),
-						'vat_amount'  => $item_tax_amount,
-						'vat'         => $item_tax_percentage,
-						'amount'      => $item_line_total_amount,
-						'line_id'     => $productId
-					);
-					array_push( $items, $item );
-				}
+			if ($this->is_redirect_payment($transaction)) {
+				$items = $this->get_items_to_capture_redirect($order);
+			} else {
+				$items = $this->get_items_to_capture_embed($order);
```

Comment:
how is items different when capturing redirect versus embedded?

Reply:
For redirect, the id is like this: `item_1`, `item_2`, and the line_id is `1`, `2`, etc.

For embed, it's set to the product_id, which makes more sense.

In a new improvement, we can align the ids